akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2AlpnSupport.scala

```diff
@@ -23,12 +27,32 @@ private[http] object Http2AlpnSupport {
   /**
    * Enables server-side Http/2 ALPN support for the given engine.
    */
-  def enableForServer(engine: SSLEngine, setChosenProtocol: String ⇒ Unit): SSLEngine = {
+  def enableForServer(engine: SSLEngine, setChosenProtocol: String ⇒ Unit): SSLEngine =
+    if (isAlpnSupportedByJDK) jdkAlpnSupport(engine, setChosenProtocol)
+    else jettyAlpnSupport(engine, setChosenProtocol)
+
+  def isAlpnSupportedByJDK: Boolean =
+    // ALPN is supported starting with JDK 9
+    JavaVersion.majorVersion >= 9
+
+  private type JDK9SSLEngine = {
+    def setHandshakeApplicationProtocolSelector(selector: BiFunction[SSLEngine, ju.List[String], String]): Unit
+  }
+  def jdkAlpnSupport(engine: SSLEngine, setChosenProtocol: String ⇒ Unit): SSLEngine = {
+    engine.asInstanceOf[JDK9SSLEngine].setHandshakeApplicationProtocolSelector { (_, protocols: ju.List[String]) ⇒
```

Comment:
Ha, cool use of structural types. I guess we can't really test this automatically, can we?

Reply:
I configured nightlies with JDK 11 here: https://jenkins.akka.io:8498/job/akka-http-nightly/

Reply:
We probably will have to revisit that code because it uses reflection which right now shows a warning:

```
[error] WARNING: An illegal reflective access operation has occurred
[error] WARNING: Illegal reflective access by scala.reflect.package$ (file:/tmp/sbt_8c5ade58/target/36b23483/scala-library-2.12.8.jar) to method sun.security.ssl.SSLEngineImpl.setHandshakeApplicationProtocolSelector(java.util.function.BiFunction)
[error] WARNING: Please consider reporting this to the maintainers of scala.reflect.package$
[error] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[error] WARNING: All illegal access operations will be denied in a future release
```

The alternative would be to somehow provide JDK 9+ binaries.